lib/data_scripts/update_bangladesh_regions_script.rb

```diff
@@ -0,0 +1,114 @@
+      logger.debug { "Processing #{row[:facility_name]}" }
+
+      facility_name = row[:facility_name]
+      division_name = row[:division]
+      district_name = row[:district]
+      upazila_name = row[:upazila]
+      facility_size = case row[:facility_type]
+        when "CC", "USC" then "community"
+        when "UHC" then "large"
+        else raise ArgumentError, "unknown facility_type #{row[:facility_type]}"
+      end
+
+      division_region = find_or_create_region(:state, division_name, org_region)
+      district_region = find_or_create_region(:district, district_name, division_region)
+      facility_group = FacilityGroup.find_by(name: district_name) || FacilityGroup.new(name: district_name, organization: org_region.source, region: district_region, protocol: protocol, generating_seed_data: true)
+      if facility_group.new_record?
+        if run_safely { facility_group.save! }
+          results[:created][:facility_groups] += 1
+        else
+          logger.warn(errors: facility_group.errors, msg: "Errors trying to save facility group #{facility_group.name}")
+          results[:errors][:facility_groups] += 1
+        end
+      end
+      upazila_region = find_or_create_region(:block, upazila_name, district_region)
+      facility_region = find_or_create_region(:facility, facility_name, upazila_region)
+      facility = Facility.new(name: facility_name, facility_group: facility_group, region: facility_region, state: division_region.name, facility_size: facility_size, zone: upazila_region.name, district: district_region.name, country: "Bangladesh")
+      if run_safely { facility.save }
+        results[:created][:facilities] += 1
+      else
+        logger.warn(errors: facility.errors, msg: "Errors trying to save facility #{facility.name}")
+        results[:errors][:facilities] += 1
+      end
+    end
+
+    results
+  end
+
+  def find_or_create_region(region_type, name, parent)
+    cache[region_type][name] ||= Region.where(region_type: region_type).find_by(name: name) || create_region(region_type, name, parent)
+  end
+
+  def create_region(region_type, name, parent)
+    region = Region.new(name: name, region_type: region_type, reparent_to: parent)
+    if run_safely { region.save }
+      results[:created][:regions] += 1
+    else
+      results[:errors][:regions] += 1
+    end
+    region
+  end
+
+  def destroy_empty_facilities
+    sql = <<-SQL
+      NOT EXISTS (SELECT 1 FROM patients where patients.registration_facility_id = facilities.id) AND
+      NOT EXISTS (SELECT 1 FROM patients where patients.assigned_facility_id = facilities.id)
+    SQL
+    facilities = Facility.where(facility_size: ["community", nil]).where(sql)
+    facilities.each do |facility|
+      if run_safely { facility.destroy }
+        results[:deleted][:facilities] += 1
+      end
+    end
+  end
+
+  def run_safely
+    return true if dry_run?
```

Comment:
Allow us to wrap persistence calls and just return true when in dry_run, to avoid accidentally updating data during a dry run.

Reply:
Neat!